File: test/plot/window.py
#! /usr/bin/env python
################################################################################
#   window.py: Plot graphs for functions in <tuna/window.h>
#
#   Copyright (C) 2014 Paul Barker
#
#   This program is free software; you can redistribute it and/or modify it
#   under the terms of the GNU General Public License as published by the
#   Free Software Foundation; either version 2, or (at your option) any
#   later version.
#
#   This program is distributed in the hope that it will be useful, but
#   WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   General Public License for more details.
################################################################################

import matplotlib
matplotlib.use('PDF')

import libtuna
import matplotlib.pyplot as plt
import numpy as np
import tuna_plot

tuna_plot.init()

length = 1000
w = np.empty([length,], dtype=np.float32)
r = libtuna.window_init_sine(w)

fig, ax = plt.subplots(figsize=(8,6), dpi=1200)
ax.plot(w)
ax.set_title("Sine window with 1000 samples")
ax.set_xlabel("Sample number")
ax.set_ylabel("Window coefficient")
ax.set_xlim(0, 1000)
#ax.set_ylim(0, 2.0)
fig.savefig("test/plot/window.pdf")

tuna_plot.exit()

File: libtuna/window.c
/*******************************************************************************
	window.c: Window function.

	Copyright (C) 2013 Paul Barker, Loughborough University
	
	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*******************************************************************************/

#include <assert.h>
#include <math.h>

#include "types.h"
#include "window.h"

void window_init_sine(float * window, uint length)
{
	uint i;

	/* We need to preserve the total energy content of the window - that is,
	 * for some constant input x[i] and our window w[i], the sum of
	 * (x[i] * w[i])^2 should be equal to the sum of (x[i])^2. The scale
	 * value is chosen to meet this condition.
	 */
	float scale = sqrtf(2.0f);

	assert(window);

	for (i = 0; i < length; i++) {
		window[i] = scale * sin(M_PI * i / length);
	}
}
